release/models/bgp/openconfig-bgp-policy.yang

```diff
@@ -661,6 +670,296 @@ module openconfig-bgp-policy {
+                  config false;
+                  uses ext-community-route-target-ip-config;
+              }
+          }
+
+          container route-origin {
+              description
+              "RFC4360 section 5 BGP extended community using an ASN.";
+
+              container config {
+                  description "Config for route origin extended community";
+                  uses ext-community-route-origin-config;
+              }
+              container state {
+                  description "State for route origin extended community";
+                  config false;
+                  uses ext-community-route-origin-config;
+              }
+          }
+
+          container route-origin-ip {
+              description
+              "RFC4360 section 5  BGP extended community using an IP
+              address.";
+
+              container config {
+                  description "Config for route origin extended
+                  community using an IP address";
+                  uses ext-community-route-origin-ip-config;
+              }
+              container state {
+                  description "State for route origin extended
+                  community using an IP address";
+                  config false;
+                  uses ext-community-route-origin-ip-config;
+              }
+          }
+
+          container color {
+              description
+              "BGP color extended community defined by RFC9012 and
+              draft-ietf-idr-segment-routing-te-policy.";
+
+              container config {
+                  description "Config for BGP color extended community
+                  defined by RFC9012";
+                  uses ext-community-color-config;
+              }
+              container state {
+                  description "State for BGP color extended community
+                  defined by RFC9012";
+                  config false;
+                  uses ext-community-color-config;
+              }
+          }
+        }
+      }
+  }
+
+grouping ext-community-asn-type-config {
+  description "Grouping for asn-type extended community config.";
+  leaf asn {
+      type oc-inet:as-number;
+      description "Autonomous system number.";
+  }
+
+  leaf value {
+      type    uint16;
+      description "Assigned value for this community.";
+  }
+}
+
+grouping ext-community-ip-type-config {
+  description "Grouping for ip-type extended community config.";
+  leaf asn {
+    type oc-inet:ip-address;
+    description "IPv4 Address for the community.";
+  }
+
+  leaf value {
+    type    uint16;
+    description "Assigned value for this community.";
+  }
+}
+
+grouping ext-community-link-bw-config {
+  description "Grouping for link bandwidth extended community config.";
+  leaf asn {
+    type oc-inet:as-number;
+    description "Autonomous system number.";
+  }
+
+  leaf bps {
+    type uint64;
+    description "Bandwidth in bits per second.";
+  }
+}
```

Comment:
The link-bandwidth draft has a number of modeling challenges.  Here's the relevant text from https://datatracker.ietf.org/doc/html/draft-ietf-idr-link-bandwidth-07

> The extended community is optional non-transitive.  The value of the
>    high-order octet of the extended Type Field is 0x40.  The value of
>    the low-order octet of the extended type field for this community is
>    0x04.  The value of the Global Administrator subfield in the Value
>    Field SHOULD represent the Autonomous System of the router that
>    attaches the Link Bandwidth Community.  If four octet AS numbering
>    scheme is used [[RFC6793](https://datatracker.ietf.org/doc/html/rfc6793)], AS_TRANS should be used in the Global
>    Administrator subfield.  The bandwidth of the link is expressed as 4
>    octets in IEEE floating point format, units being bytes (not bits!)
>    per second.  It is carried in the Local Administrator subfield of the
>    Value Field.

Thus, for config purposes you *could* use an as-number type, but it'll be reduced into AS_TRANS by the implementations if it's following the draft.  

Juniper's implementation simply wants a 16-bit number there.

The "bps" leaf is also at odds with the contents of the draft being *bytes* per second.  I suggest addressing that.

The place where things are a significant headache is that the protocol component is carried in floating point format on the wire.  While Juniper's implementation limits the input to a 32-bit integer, a 64-bit integer could eventually be implemented - but it's incompatible from original commit if so.  But regardless of whether the input is 32 or 64-bit at configuration, eventually on the wire it is reduced to a IEEE 32-bit float and the number is coerced into something that no longer matches the input.

Implementations thus often take the integer input, and in operations, display some form of float output or a integer truncated output for the on-the-wire float.

Reply:
Updated to use float32 for the bandwidth as suggested.

I will remove the ASN leaf for link bandwidth since it essentially cannot be configured.